src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+                        throw WinHttpException.CreateExceptionUsingError(lastError);
+                    }
+                }
+
+                // Re-add cookies. The GetCookieHeader() method will return the correct set of
+                // cookies based on the redirectUri.
+                string cookieHeader = WinHttpHandler.GetCookieHeader(redirectUri, state.Handler.CookieContainer);
+                if (!string.IsNullOrEmpty(cookieHeader))
+                {
+                    if (!Interop.WinHttp.WinHttpAddRequestHeaders(
+                        state.RequestHandle,
+                        cookieHeader,
+                        (uint)cookieHeader.Length,
+                        Interop.WinHttp.WINHTTP_ADDREQ_FLAG_ADD))
+                    {
+                        WinHttpException.ThrowExceptionUsingLastError();
+                    }
+                }
+            }
+
+            state.RequestMessage.RequestUri = redirectUri;
+            
+            // Redirection to a new uri may require a new connection through a potentially different proxy.
+            // If so, we will need to respond to additional 407 proxy auth demands and re-attach any
+            // proxy credentials. The ProcessResponse() method looks at the state.LastStatusCode
+            // before attaching proxy credentials and marking the HTTP request to be re-submitted.
+            // So we need to reset the LastStatusCode remembered. Otherwise, it will see additional 407
+            // responses as an indication that proxy auth failed and won't retry the HTTP request.
+            if (state.LastStatusCode == HttpStatusCode.ProxyAuthenticationRequired)
+            {
+                state.LastStatusCode = 0;
+            }
+
+            // For security reasons, we drop the server credential if it is a 
+            // NetworkCredential.  But we allow credentials in a CredentialCache
+            // since they are specifically tied to URI's.
+            if (!(state.ServerCredentials is CredentialCache))
+            {
+                state.ServerCredentials = null;
+            }
+        }
+        
+        private static void OnRequestSendingRequest(WinHttpRequestState state)
+        {
+            Debug.Assert(state != null, "OnRequestSendingRequest: state is null");
+            
+            if (state.RequestMessage.RequestUri.Scheme != WinHttpRequestState.UriSchemeHttps)
+            {
+                // Not SSL/TLS.
+                return;
+            }
+
+            // Grab the channel binding token (CBT) information from the request handle and put it into
+            // the TransportContext object.
+            state.TransportContext.SetChannelBinding(state.RequestHandle);
+
+            if (state.ServerCertificateValidationCallback != null)
+            {
+                IntPtr certHandle = IntPtr.Zero;
+                uint certHandleSize = (uint)IntPtr.Size;
+
+                if (!Interop.WinHttp.WinHttpQueryOption(
```

Comment:
Missing `lock(state.Lock)` - no two WinHttp operations using the _same_ handle should occur in parallel.
(Not adding these comments but they do apply to this file and potentially other places where this situation could occur via either lib or user code.)